### PR TITLE
Implement Deserialize and Serialize

### PR DESCRIPTION
**Pull request by chinedufn**
## What

This is a work in progress PR for beginning to allow `rust-protobuf` enums to implement Serialize / Deserialize from `serde`.

*EDIT*

Adding support for messages as well, not just enums.

Related issue: #242

### CODE REVIEW

**Comment by chinedufn** [comment]
Hey @stepancheg,

You mentioned that you'd be able to take a look at a WIP PR so I thought I'd fire off my first question here:

I'm unable to run my test since I haven't yet figured out how to enable the `with-serde` feature.

My first thought was to verify that I can run the `with-bytes` tests and then track down what was different about my `with-serde` feature setup.

Unfortunately - I couldn't run the `with-bytes` tests either. I must be doing something wrong here:

```
$ cargo test --features=with-bytes --package protobuf-test --lib common::v2::test_carllerche_bytes::test -- --nocapture
    Finished dev [unoptimized + debuginfo] target(s) in 0.12s
     Running target/debug/deps/protobuf_test-927a428e5dae8e14

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 145 filtered out
```

```
$ cargo -V
cargo 1.28.0-nightly (f352115d5 2018-05-15)
```

Anything stick out?

**Comment by stepancheg** [comment]
A couple of questions at the moment.

You are aiming to implementing `derive` for enums only? What for?

Protobuf crate doesn't depend on serde. Thus `with-serde` option AFAIU is not needed for `protobuf`.

**Comment by chinedufn** [comment]
> You are aiming to implementing derive for enums only? What is the use case?

My goal here was to start with enums because I thought that they didn't have any fields that needed to be ignored / renamed / etc. Just straight up `#derive(Serialize, Deserialize)`.

So I more or less wanted to quickly get something working then they struct implementation could come in a separate PR.

For all I know that approach isn't actually easier / doesn't make sense so would appreciate your feedback here.

> Protobuf crate doesn't depend on serde. Thus with-serde option AFAIU is not needed for protobuf.

Great thanks I've removed serde from `protobuf/Cargo.toml`

**Comment by stepancheg** [comment]
If you are just starting with enums, that's fine. I'd just call an option `serde_derive`, not `serde_derive_for_enums` to avoid confusion.

But generally, what's your use case of serde for protobuf?

**Comment by stepancheg** [comment]
What else.

As protobuf no longer depends on serde, I think `serde` option is no longer needed in tests: `protobuf-test` crate can depend on serde unconditionally, and it's unlikely that something will break when serde is unavailable.

**Comment by chinedufn** [comment]
> As protobuf no longer depends on serde, I think serde option is no longer needed in tests: protobuf-test crate can depend on serde unconditionally, and it's unlikely that something will break when serde is unavailable.

Addressed here - 9b4ae83 lmk if I missed anything!

> But generally, what's your use case of serde for protobuf?

I'm working on a game that has an `Entity` struct. An `Entity` can have a `model` (among many other things...)

```
#[derive(Deserialize)]
struct Entity {
  model: MySerdeEnumHere
}
```

And then I have files that specify default entities for my games and some of their properties and I deserialize those into `Vec<Entity>` when the game starts.

---

Right now I'm running a bash script to add `Deserialize` too all of my rust-protobuf enums

```
# ...
  sed -i '' -e 's/Clone,PartialEq,Eq/Clone,Deserialize,PartialEq,Eq/g' $file
# ...
```

Which works fine..

But if rust-protobuf supported serde I wouldn't need this hack.

File: protobuf-test/README.md [hunk 1]
```diff
@@ -7,6 +7,9 @@ If protobuf 3 is installed, command will include test for protobuf 3 generated c
 If `--features=with-bytes` flag is specified, tests will include test for `with-bytes` feature,
 which is not enabled by default.
 
+If `--features=with-serde` flag is specified, tests will include test for `with-serde` feature,
```
Comment (stepancheg) [inline protobuf-test/README.md]: This comment is outdated now.

**Comment by stepancheg** [comment]
> And then I have files that specify default entities for my games and some of their properties and I deserialize those into Vec<Entity> when the game starts.

Which format are you using for serialization? JSON, some other binary?

**Comment by chinedufn** [comment]
I'm use `JSON` and [ron](https://github.com/ron-rs/ron) at the moment but might add other formats in the future

**Comment by stepancheg** [comment]
OK, as I said, rust-protobuf should support JSON encoding natively, but if `rod` is needed (and I [started implementing it](https://github.com/stepancheg/rust-protobuf/tree/master/protobuf/src/json)), but if `ron` and other formats needed, that's fine.

Next thing you need to do is add `serde_derive` option to `Customize` struct.

File: protobuf-test-common/Cargo.toml [hunk 2]
```diff
@@ -15,6 +15,7 @@ bench = false
 default-features = []
 proto3 = []
 with-bytes = ["bytes", "protobuf/with-bytes"]
+with-serde = ["serde", "serde_derive"]
```
Comment (stepancheg) [inline protobuf-test-common/Cargo.toml]: This option is not needed here either.

**Comment by chinedufn** [comment]
Alright I added the `serde_derive` option to Customize and I've decided to support messages/structs in this PR and not just enums now that I'm feeling a tad more comfy (thanks to all of your help!)

Next up I'm going to look at the `protobuf-codegen` crate to find out where I'll need to conditionally add `Serialize, Deserialize` ...

Let me know if you have a better approach in mind!

**Comment by stepancheg** [comment]
`derive` for messages is added in `protobuf-codegen/src/message.rs`, `write_struct` function.

`MessageGen` type already has `customize` field where your new `serde_derive` option should be.

`derive` for enums is added in `protobuf-codegen/src/enums.rs`, `write_struct` function.

`EnumGen` doesn't have `Customize`, it should be added there I think.

**Comment by chinedufn** [comment]
Done! Thank you so much for your guidance!

Next I need to generate `[serde(ignore)]` above cached size and unknown fields

File: protobuf-codegen/src/message.rs [hunk 3]
```diff
@@ -347,6 +347,10 @@ impl<'a> MessageGen<'a> {
         self.fields.len() <= 500
     }
 
+    fn supports_serde(&self) -> bool {
+        self.customize.serde_derive.is_some() && self.customize.serde_derive.unwrap()
```
Comment (stepancheg) [inline protobuf-codegen/src/message.rs]: This could be written as `self.customize.serde_derive.unwrap_or(false)`.

Could you please inline this function `supports_serde` and put this condition near `derive` generation?

File: protobuf-test-common/Cargo.toml [hunk 4]
```diff
@@ -21,6 +21,8 @@ glob        = "0.2"
 log         = "0.*"
 env_logger  = "0.5.*"
 tempdir     = "~0.3"
+serde = "1.0"
```
Comment (stepancheg) [inline protobuf-test-common/Cargo.toml]: Align spaces as in previous dependencies, please?

**Comment by stepancheg** [comment]
Also please note there are missing changes in `protobuf-codegen-pure-test`: it should change `Cargo.toml` similarly to `Cargo.toml` in protobuf-test.

File: protobuf-codegen/src/message.rs (hunk 3 above)
Comment (chinedufn) [reply protobuf-codegen/src/message.rs]: Inlined it - but now I'm realizing that I'll need to call this again when conditionally adding `#[serde(ignore)]` so I'm thinking we keep this in a function?

Comment (stepancheg) [reply protobuf-codegen/src/message.rs]: OK, but call it something different then, like `serde_derive_enabled` or `enabled_serde_derive` then.

`supports_derive_partial_eq` is different, it is not about querying options, it is about work around rustc bug.

Comment (chinedufn) [reply protobuf-codegen/src/message.rs]: Awesome - thanks - renamed

File: protobuf-codegen/src/message.rs [hunk 5]
```diff
@@ -356,6 +358,9 @@ impl<'a> MessageGen<'a> {
         if self.lite_runtime {
             derive.push("Debug");
         }
+        if self.serde_derive_enabled() {
+            derive.push("Serialize,Deserialize");
```
Comment (stepancheg) [inline protobuf-codegen/src/message.rs]: Should be

```
derive.extend(&["Serialize", "Deserialize"]);
```

**Comment by stepancheg** [comment]
For the first version, some tests need to be added. Just to verify at least something is working.

In proper version there should be more tests (which will break code), like:
* compatibility with `with-bytes` feature
* map fields
* oneof fields

File: protobuf-test/src/common/v2/test_serde_derive.rs [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+#[test]
+fn serialize () {
+    let mut data = TestSerde::new();
+    data.set_test(SerdeEnum::TEST);
+
+    let serialized = serde_json::to_string(&data).unwrap();
+
+    assert_eq!(serialized, "{\"test\":\"TEST\"}");
+    println!("{:#?}", serialized);
```
Comment (stepancheg) [inline protobuf-test/src/common/v2/test_serde_derive.rs]: No need to print anything in tests.

File: protobuf-test/src/common/v2/test_serde_derive.rs [hunk 7]
```diff
@@ -0,0 +1,17 @@
+use super::test_serde_derive_pb::*;
+use serde_json;
+
+#[derive(Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+struct Proto(TestSerde);
```
Comment (stepancheg) [inline protobuf-test/src/common/v2/test_serde_derive.rs]: Not clear how `struct Proto` is part of the test.

File: protobuf-test/src/common/v2/test_serde_derive.rs (hunk 6 above)
Comment (chinedufn) [reply protobuf-test/src/common/v2/test_serde_derive.rs]: 👍 accident - cleaning up now

File: protobuf-test/src/common/v2/test_serde_derive.rs (hunk 7 above)
Comment (chinedufn) [reply protobuf-test/src/common/v2/test_serde_derive.rs]: I was originally "testing" by just verifying that we could implement Serialize and Deserialize (it would fail if TestSerde wasn't serializable.. but now I'm going with a more explicit test function.

File: protobuf-codegen/src/enums.rs [hunk 8]
```diff
@@ -145,6 +147,9 @@ impl<'a> EnumGen<'a> {
                 "Note: you cannot use pattern matching for enums with allow_alias option",
             );
         }
+        if self.customize.serde_derive.unwrap_or(false) {
+            derive.push("Serialize,Deserialize");
```
Comment (stepancheg) [inline protobuf-codegen/src/enums.rs]:
```
derive.extend(&["Serialize,Deserialize"]);
```

File: protobuf-test/src/common/v2/test_serde_derive.rs (hunk 7 above)
Comment (chinedufn) [reply protobuf-test/src/common/v2/test_serde_derive.rs]: I was originally "testing" by just verifying that we could implement Serialize and Deserialize (it would fail if TestSerde wasn't serializable.. but now I'm going with a more explicit test function.

File: README.md [hunk 9]
```diff
@@ -103,6 +103,50 @@ generated as `Bytes` or `Chars` respectively. When `CodedInputStream` is constru
...
+
+Rust-protobuf can be used with [serde](https://github.com/serde-rs/serde).
+
+To enable `serde` you need to:
+
+1. Enable `with-serde` feature in rust-protobuf:
+
+```
+[dependencies]
+protobuf = { version = "~2.0", features = ["with-serde"] }
```
Comment (stepancheg) [inline README.md]: This is outdated.

File: README.md [hunk 10]
```diff
@@ -103,6 +103,50 @@ generated as `Bytes` or `Chars` respectively. When `CodedInputStream` is constru
...
+```
+
+2. Enable serde option
+
+with Customize when codegen is invoked programmatically:
```
Comment (stepancheg) [inline README.md]: with `Customize` when...

File: README.md [hunk 11]
```diff
@@ -103,6 +103,50 @@ generated as `Bytes` or `Chars` respectively. When `CodedInputStream` is constru
...
+import "rustproto.proto";
+
+option (rustproto.serde_derive_all) = true;
+```
+
+You may now `Serialize` and `Deserialize` messages:
+
+```rs
+#[derive(Serialize, Deserialize)]
+struct MyStruct {
```
Comment (stepancheg) [inline README.md]: This is a bit misleading: developer doesn't need to put anything into another struct to serialize/deserialize.

File: protobuf-test/src/common/v2/test_serde_derive.rs (hunk 6 above)
Comment (stepancheg) [reply protobuf-test/src/common/v2/test_serde_derive.rs]: Would be hard to add deserialization test here?

**Comment by chinedufn** [comment]
Adding tests for `oneof` and `map` in this PR. Not sure how to test against `with-bytes` - but you mentioned it isn't mandatory for v1 so I'll get this `oneof` test passing.

**Comment by chinedufn** [comment]
Also adding a test for `repeated` .. Any guidance on where to look to get these passing would be appreciated but in the meantime I can take a look at the failing test error message and start digging around from there.

**Comment by stepancheg** [comment]
OK, I can merge this PR now. Tell if you are OK with it.

**Comment by chinedufn** [comment]
Here are the errors for the one of test (haven't started making it pass). I'm going to go grab some food then take a look but plopping them here in case you know the obvious places to make code changes.

```
error[E0277]: the trait bound `protobuf::SingularPtrField<common::v2::test_serde_derive_pb::SerdeOneOf>: serde::Serialize` is not satisfied
  --> protobuf-test/src/common/v2/test_serde_derive_pb.rs:24:34
   |
24 | #[derive(PartialEq,Clone,Default,Serialize,Deserialize)]
   |                                  ^^^^^^^^^ the trait `serde::Serialize` is not implemented for `protobuf::SingularPtrField<common::v2::test_serde_derive_pb::SerdeOneOf>`
   |
   = note: required by `serde::ser::SerializeStruct::serialize_field`

error[E0277]: the trait bound `protobuf::SingularPtrField<common::v2::test_serde_derive_pb::SerdeOneOf>: serde::Deserialize<'_>` is not satisfied
  --> protobuf-test/src/common/v2/test_serde_derive_pb.rs:24:44
   |
24 | #[derive(PartialEq,Clone,Default,Serialize,Deserialize)]
   |                                            ^^^^^^^^^^^ the trait `serde::Deserialize<'_>` is not implemented for `protobuf::SingularPtrField<common::v2::test_serde_derive_pb::SerdeOneOf>`
   |
   = note: required by `serde::de::SeqAccess::next_element`

error[E0277]: the trait bound `protobuf::SingularPtrField<common::v2::test_serde_derive_pb::SerdeOneOf>: serde::Deserialize<'_>` is not satisfied
  --> protobuf-test/src/common/v2/test_serde_derive_pb.rs:24:44
   |
24 | #[derive(PartialEq,Clone,Default,Serialize,Deserialize)]
   |                                            ^^^^^^^^^^^ the trait `serde::Deserialize<'_>` is not implemented for `protobuf::SingularPtrField<common::v2::test_serde_derive_pb::SerdeOneOf>`
   |
   = note: required by `serde::de::MapAccess::next_value`

error[E0277]: the trait bound `common::v2::test_serde_derive_pb::SerdeOneOf_oneof_food: serde::Serialize` is not satisfied
   --> protobuf-test/src/common/v2/test_serde_derive_pb.rs:305:34
    |
305 | #[derive(PartialEq,Clone,Default,Serialize,Deserialize)]
    |                                  ^^^^^^^^^ the trait `serde::Serialize` is not implemented for `common::v2::test_serde_derive_pb::SerdeOneOf_oneof_food`
    |
    = note: required because of the requirements on the impl of `serde::Serialize` for `std::option::Option<common::v2::test_serde_derive_pb::SerdeOneOf_oneof_food>`
    = note: required by `serde::ser::SerializeStruct::serialize_field`

error[E0277]: the trait bound `common::v2::test_serde_derive_pb::SerdeOneOf_oneof_food: serde::Deserialize<'_>` is not satisfied
   --> protobuf-test/src/common/v2/test_serde_derive_pb.rs:305:44
    |
305 | #[derive(PartialEq,Clone,Default,Serialize,Deserialize)]
    |                                            ^^^^^^^^^^^ the trait `serde::Deserialize<'_>` is not implemented for `common::v2::test_serde_derive_pb::SerdeOneOf_oneof_food`
    |
    = note: required because of the requirements on the impl of `serde::Deserialize<'_>` for `std::option::Option<common::v2::test_serde_derive_pb::SerdeOneOf_oneof_food>`
    = note: required by `serde::de::SeqAccess::next_element`

error[E0277]: the trait bound `common::v2::test_serde_derive_pb::SerdeOneOf_oneof_food: serde::Deserialize<'_>` is not satisfied
   --> protobuf-test/src/common/v2/test_serde_derive_pb.rs:305:44
    |
305 | #[derive(PartialEq,Clone,Default,Serialize,Deserialize)]
    |                                            ^^^^^^^^^^^ the trait `serde::Deserialize<'_>` is not implemented for `common::v2::test_serde_derive_pb::SerdeOneOf_oneof_food`
    |
    = note: required because of the requirements on the impl of `serde::Deserialize<'_>` for `std::option::Option<common::v2::test_serde_derive_pb::SerdeOneOf_oneof_food>`
    = note: required by `serde::de::MapAccess::next_value`

error[E0277]: the trait bound `protobuf::SingularPtrField<common::v3::test_serde_derive_pb::SerdeOneOf>: serde::Serialize` is not satisfied
  --> protobuf-test/src/common/v3/test_serde_derive_pb.rs:24:34
   |
24 | #[derive(PartialEq,Clone,Default,Serialize,Deserialize)]
   |                                  ^^^^^^^^^ the trait `serde::Serialize` is not implemented for `protobuf::SingularPtrField<common::v3::test_serde_derive_pb::SerdeOneOf>`
   |
   = note: required by `serde::ser::SerializeStruct::serialize_field`

error[E0277]: the trait bound `protobuf::SingularPtrField<common::v3::test_serde_derive_pb::SerdeOneOf>: serde::Deserialize<'_>` is not satisfied
  --> protobuf-test/src/common/v3/test_serde_derive_pb.rs:24:44
   |
24 | #[derive(PartialEq,Clone,Default,Serialize,Deserialize)]
   |                                            ^^^^^^^^^^^ the trait `serde::Deserialize<'_>` is not implemented for `protobuf::SingularPtrField<common::v3::test_serde_derive_pb::SerdeOneOf>`
   |
   = note: required by `serde::de::SeqAccess::next_element`

error[E0277]: the trait bound `protobuf::SingularPtrField<common::v3::test_serde_derive_pb::SerdeOneOf>: serde::Deserialize<'_>` is not satisfied
  --> protobuf-test/src/common/v3/test_serde_derive_pb.rs:24:44
   |
24 | #[derive(PartialEq,Clone,Default,Serialize,Deserialize)]
   |                                            ^^^^^^^^^^^ the trait `serde::Deserialize<'_>` is not implemented for `protobuf::SingularPtrField<common::v3::test_serde_derive_pb::SerdeOneOf>`
   |
   = note: required by `serde::de::MapAccess::next_value`

error[E0277]: the trait bound `common::v3::test_serde_derive_pb::SerdeOneOf_oneof_food: serde::Serialize` is not satisfied
   --> protobuf-test/src/common/v3/test_serde_derive_pb.rs:300:34
    |
300 | #[derive(PartialEq,Clone,Default,Serialize,Deserialize)]
    |                                  ^^^^^^^^^ the trait `serde::Serialize` is not implemented for `common::v3::test_serde_derive_pb::SerdeOneOf_oneof_food`
    |
    = note: required because of the requirements on the impl of `serde::Serialize` for `std::option::Option<common::v3::test_serde_derive_pb::SerdeOneOf_oneof_food>`
    = note: required by `serde::ser::SerializeStruct::serialize_field`

error[E0277]: the trait bound `common::v3::test_serde_derive_pb::SerdeOneOf_oneof_food: serde::Deserialize<'_>` is not satisfied
   --> protobuf-test/src/common/v3/test_serde_derive_pb.rs:300:44
    |
300 | #[derive(PartialEq,Clone,Default,Serialize,Deserialize)]
    |                                            ^^^^^^^^^^^ the trait `serde::Deserialize<'_>` is not implemented for `common::v3::test_serde_derive_pb::SerdeOneOf_oneof_food`
    |
    = note: required because of the requirements on the impl of `serde::Deserialize<'_>` for `std::option::Option<common::v3::test_serde_derive_pb::SerdeOneOf_oneof_food>`
    = note: required by `serde::de::SeqAccess::next_element`

error[E0277]: the trait bound `common::v3::test_serde_derive_pb::SerdeOneOf_oneof_food: serde::Deserialize<'_>` is not satisfied
   --> protobuf-test/src/common/v3/test_serde_derive_pb.rs:300:44
    |
300 | #[derive(PartialEq,Clone,Default,Serialize,Deserialize)]
    |                                            ^^^^^^^^^^^ the trait `serde::Deserialize<'_>` is not implemented for `common::v3::test_serde_derive_pb::SerdeOneOf_oneof_food`
    |
    = note: required because of the requirements on the impl of `serde::Deserialize<'_>` for `std::option::Option<common::v3::test_serde_derive_pb::SerdeOneOf_oneof_food>`
    = note: required by `serde::de::MapAccess::next_value`

error: aborting due to 12 previous errors

For more information about this error, try `rustc --explain E0277`.
error: Could not compile `protobuf-test`.

To learn more, run the command again with --verbose.

Process finished with exit code 101
```

**Comment by chinedufn** [comment]
I'd like to get the `test_oneof` and `repeated` test passing then merge - if that's alright with you?

**Comment by stepancheg** [comment]
```
SerdeOneOf_oneof_food
```

Derive for `oneof` should be added. Somewhere in `oneof.rs`.

```
SingularPtrField
```

This one is harder to fix.

I think it should be custom serialization for this field in a separate crate like `protobuf-serde` and fields should be probably annotated with 

```
#[serde(deserialize_with = "::protobuf-serde::...")]
```

However, temporarily in the test this could be turned off with `repeated_field_vec` option set to `true`.

**Comment by stepancheg** [comment]
> I'd like to get the test_oneof and repeated test passing then merge - if that's alright with you?

Yep.

**Comment by chinedufn** [comment]
So the `SingularPtrField` lives in the `protobuf` crate right now, which does not have `serde`.

Do I need to add the `with-serde` back so that we can conditionally add the `serde(deserialize_with` ?

**Comment by stepancheg** [comment]
As I said, serialization can be placed in separate crate `protobuf-serde`, and that crate may be referenced with

```
#[serde(deserialize_with = "::protobuf-serde::...")]
```

But for `RepeatedPtrField` test could be temporarily disabled with option `repeated_field_vec = true`. For singular field similar option can be implemented like `singular_field_option` (I forgot to do it). I guess it will be slightly easier.

You choose.

**Comment by stepancheg** [comment]
OK, now I doubt that implementing `singular_field_option` is easier.

I've created an issue anyway: https://github.com/stepancheg/rust-protobuf/issues/300

File: protobuf-codegen/src/oneof.rs [hunk 12]
```diff
@@ -156,6 +156,9 @@ impl<'a> OneofGen<'a> {
         if self.lite_runtime {
             derive.push("Debug");
         }
+        if self.customize.serde_derive.unwrap_or(false) {
+            derive.extend(&["Serialize,Deserialize"]);
```
Comment (stepancheg) [inline protobuf-codegen/src/oneof.rs]: Should be

```
derive.extend(&["Serialize", "Deserialize"]); 
```

File: protobuf-test/src/common/v2/test_serde_derive.rs [hunk 13]
```diff
@@ -0,0 +1,29 @@
+use super::test_serde_derive_pb::*;
+use serde_json;
+use std::collections::HashMap;
+
+#[test]
+fn serialize_deserialize () {
```
Comment (stepancheg) [inline protobuf-test/src/common/v2/test_serde_derive.rs]: Could you please split this test into smaller tests, e. g. `test_enum`, `test_map`, `test_oneof` etc.

File: protobuf-codegen/src/enums.rs [hunk 14]
```diff
@@ -145,6 +147,9 @@ impl<'a> EnumGen<'a> {
                 "Note: you cannot use pattern matching for enums with allow_alias option",
             );
         }
+        if self.customize.serde_derive.unwrap_or(false) {
+            derive.extend(&["Serialize,Deserialize"]);
```
Comment (stepancheg) [inline protobuf-codegen/src/enums.rs]:
```
derive.extend(&["Serialize", "Deserialize"]); 
```

File: protobuf-serde/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,5 @@
+pub fn serialize_oneof () {
```
Comment (stepancheg) [inline protobuf-serde/src/lib.rs]: I assume this is just a stub yet which you are going to implement. Could you please add TODO next time to avoid confusion?

BTW, `oneof` can be serialized with `derive`. Custom serializers are needed for `RepeatedField` etc.

Comment (chinedufn) [reply protobuf-serde/src/lib.rs]: If we serialize a `OneOf` it looks like it'll include the `value` and `set` fields whereas we just want the `value` so I'm thinking we'll want a custom serializer here

So that we end up with

```
{ "oneOfField": {"datatype": {}} } 
```

Instead of

```
{"oneOfField": {"value": {"datatype": {}", "set": true}}}
```

File: protobuf-test/src/common/v2/test_serde_derive.rs (hunk 13 above)
Comment (chinedufn) [reply protobuf-test/src/common/v2/test_serde_derive.rs]: Yup - mind if I split it up once I have everything working? Or is it hurting your reviewing at the moment (if so I'll split it up right now, no big deal)

File: protobuf-serde/src/lib.rs (hunk 15 above)
Comment (stepancheg) [reply protobuf-serde/src/lib.rs]: > `{"oneOfField": {"value": {"datatype": {}", "set": true}}}`

Is it serde adding `set` field? Why?

What is `datatype`? Could you please post a protobuf schema for this JSON generated, because it's hard to reply without understanding context?

Anyway

* it's not possible to implement oneof serializer in `protobuf-serde`, protobuf type system won't allow it
* this is exactly the reason why I think JSON mapping should be implemented natively

Comment (stepancheg) [reply protobuf-serde/src/lib.rs]: Now I get it. It's singular field (`optional` or `repeated`), not `oneof`.

Comment (chinedufn) [reply protobuf-serde/src/lib.rs]: Ah apologies for the confusion - yeah it's this:

https://github.com/stepancheg/rust-protobuf/blob/d153996e9354abe3c99933a3224288064056ae30/protobuf-test/src/common/v2/test_serde_derive_pb.proto#L10

Guess I'll need to add a test for an inlined `oneof` also after I get this working.

File: protobuf-codegen/src/message.rs [hunk 16]
```diff
@@ -347,6 +347,8 @@ impl<'a> MessageGen<'a> {
         self.fields.len() <= 500
     }
 
+    fn serde_derive_enabled(&self) -> bool { self.customize.serde_derive.unwrap_or(false) }
```
Comment (stepancheg) [inline protobuf-codegen/src/message.rs]: Split into 3 lines please to match the style of the project.

File: protobuf-codegen/src/message.rs [hunk 17]
```diff
@@ -379,10 +384,17 @@ impl<'a> MessageGen<'a> {
                                 FieldKind::Oneof(..) => unreachable!(),
                             }
                         };
+
+                        let field_type = &field.full_storage_type().to_string();
+                        if self.serde_derive_enabled() && field_type.contains("::protobuf::SingularPtrField<") {
```
Comment (stepancheg) [inline protobuf-codegen/src/message.rs]: This should be a match of field kind, not on string, e. g.

```
                    if let FieldKind::Singular(singular) = field.kind {
                        if let SingularFieldFlag::WithFlag { option_kind, .. } = singular.flag {
                            if option_kind == OptionKind::SingularField {
                                // write custom serializer
                            }
                        }
                    }
```

Comment (chinedufn) [reply protobuf-codegen/src/message.rs]: Thanks for all of this instruction!

**Comment by stepancheg** [comment]
I started merging basic parts of this PR. Please rebase.

**Comment by chinedufn** [comment]
Cool - I'm just about wrapped up with the tests.

**Comment by chinedufn** [comment]
Alright I've rebased and split up the tests

**Comment by chinedufn** [comment]
Err apologies I misspoke - rebasing now..

**Comment by chinedufn** [comment]
Alright I think I'm good to go here - happy to address any additional feedback that you might have

**Comment by stepancheg** [comment]
OK, now waiting for a travis build to finish.

File: protobuf-serde/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+                                                         -> Result<<S as Serializer>::Ok, <S as Serializer>::Error>
+    where S: Serializer {
+    spf.get_ref().serialize(serializer)
+}
+
+pub fn deserialize_singular_ptr_field<'de, D, T>(deserializer: D)
+                                                 -> Result<SingularPtrField<T>, D::Error>
+    where D: Deserializer<'de>, T: Deserialize<'de>
+{
+    Ok(SingularPtrField::some(Option::deserialize(deserializer).unwrap().unwrap()))
```
Comment (stepancheg) [inline protobuf-serde/src/lib.rs]: This unwraps must be mistakes.

If it is there temporarlily, please add TODO.

Comment (stepancheg) [reply protobuf-serde/src/lib.rs]: If unwrap is the right thing, it deserved an explanation in comments at least.

File: protobuf-test/src/common/v2/test_serde_derive_pb.proto [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+message SomeMessage {
+}
+
+message Repeated {
+    repeated uint32 test_repeated = 1;
+}
+
+message TestSerdeMap {
+    map<uint32, uint32> test_map = 1;
+}
```
Comment (stepancheg) [inline protobuf-test/src/common/v2/test_serde_derive_pb.proto]: Missing trailing newline.

File: protobuf-codegen/Cargo.toml [hunk 20]
```diff
@@ -17,7 +17,7 @@ See `protoc-rust` and `protobuf-codegen-pure` crates.
 bench = false
 
 [dependencies]
-protobuf = { path = "../protobuf", version = "=3.0.0" }
+protobuf             = { path = "../protobuf", version = "=3.0.0" }
```
Comment (stepancheg) [inline protobuf-codegen/Cargo.toml]: This is strange indentation.

File: protobuf-codegen/src/message.rs [hunk 21]
```diff
@@ -379,10 +386,24 @@ impl<'a> MessageGen<'a> {
...
+                                    }
+                                }
+                            }
+                        }
+
                         w.field_decl_vis(
                             vis,
                             &field.rust_name,
-                            &field.full_storage_type().to_string(),
+                            &field_type.to_string(),
```
Comment (stepancheg) [inline protobuf-codegen/src/message.rs]: This change is unrelated to serde change, should be reverted.

File: protobuf-serde/src/lib.rs (hunk 18 above)
Comment (chinedufn) [reply protobuf-serde/src/lib.rs]: Ah ok I understand now - adding another test as we speak.

**Comment by stepancheg** [comment]
Merged another change — test diagnostics — it's unrelated to this PR. Please rebase.

File: protobuf-serde/src/lib.rs (hunk 18 above)
Comment (stepancheg) [reply protobuf-serde/src/lib.rs]: > Ah ok I understand now - adding another test as we speak.

No need to add all possible tests right now. I'm just noticed `unwrap` so I'm not convinced it's safe, so I'm saying that either:
* code should be rewritten without unwrap
* or code should have comment explanations why the panic is not possible

**Comment by chinedufn** [comment]
Rebased and all tests passing locally - just waiting on CircleCI now.

Happy to fix any other feedback that you might have.

**Comment by chinedufn** [comment]
@stepancheg alright it looks like we’re passing in CI as well - thanks a lot for the help! Happy to fix anything else that is blocking this from being merged - if anything!

**Comment by stepancheg** [comment]
Done! Thank you!

**Comment by stepancheg** [comment]
I've added checklist in the issue https://github.com/stepancheg/rust-protobuf/issues/266 tracking what needs to be done.
